src/conditions/datadog_search.rs

```diff
@@ -27,215 +26,297 @@ impl_generate_config_from_default!(DatadogSearchConfig);
         let node = parse(&self.source)?;
-        let matcher = as_log(build_matcher(&node, &EventFilter::default()));
+        let matcher = fast_matcher::build_matcher(&node, &EventFilter::default());
 
         Ok(Condition::DatadogSearch(DatadogSearchRunner { matcher }))
     }
 }
 
-/// Run the provided `Matcher` when we're dealing with `LogEvent`s. Otherwise, return false.
-fn as_log(matcher: Box<dyn Matcher<LogEvent>>) -> Box<dyn Matcher<Event>> {
-    Run::boxed(move |ev| match ev {
-        Event::Log(log) => matcher.run(log),
-        _ => false,
-    })
-}
-
 //------------------------------------------------------------------------------
 
 #[derive(Default, Clone)]
-struct EventFilter;
+pub struct EventFilter;
 
-/// Uses the default `Resolver`, to build a `Vec<Field>`.
-impl Resolver for EventFilter {}
-
-impl Filter<LogEvent> for EventFilter {
-    fn exists(&self, field: Field) -> Box<dyn Matcher<LogEvent>> {
-        match field {
-            Field::Tag(tag) => {
-                let starts_with = format!("{}:", tag);
-
-                any_string_match("tags", move |value| {
-                    value == tag || value.starts_with(&starts_with)
-                })
-            }
-            // Literal field 'tags' needs to be compared by key.
-            Field::Reserved(field) if field == "tags" => {
-                any_string_match("tags", move |value| value == field)
-            }
-            Field::Default(f) | Field::Facet(f) | Field::Reserved(f) => {
-                Run::boxed(move |log: &LogEvent| log.get(&f).is_some())
-            }
+impl EventFilter {
+    pub fn run(matcher: &fast_matcher::FastMatcher, log: &LogEvent) -> bool {
+        match &matcher.mode {
+            Mode::One(op) => exec(op, log),
+            Mode::Any(ops) => ops.iter().any(|op| exec(op, log)),
+            Mode::All(ops) => ops.iter().all(|op| exec(op, log)),
         }
     }
+}
 
-    fn equals(&self, field: Field, to_match: &str) -> Box<dyn Matcher<LogEvent>> {
-        match field {
-            // Default fields are compared by word boundary.
-            Field::Default(field) => {
-                let re = word_regex(to_match);
-
-                string_match(&field, move |value| re.is_match(&value))
-            }
-            // A literal "tags" field should match by key.
-            Field::Reserved(field) if field == "tags" => {
-                let to_match = to_match.to_owned();
+fn exec(op: &Op, log: &LogEvent) -> bool {
+    match op {
+        Op::True => true,
+        Op::False => true,
+        Op::Exists(field) => exists(field, log),
+        Op::NotExists(field) => !exists(field, log),
+        Op::Equals { field, value } => match field {
+            Field::Reserved(f) | Field::Facet(f) => equals(f, value, log),
+            _ => false,
+        },
+        Op::TagExists(value) => tag_exists(value, log),
+        Op::RegexMatch { field, re } => match field {
+            Field::Default(f) => regex_match(f, re, log),
```

Comment:
Just curious: was it not possible to coalesce this method between the `match_datadog_query` VRL function and this condition?

I know that a VRL value can be different from the more fixed layout of `LogEvent`, but it seems like -- and this is just from a cursory glance, to be fair -- that it would end being more about a simple translation layer to "query" the operand in a generic way i.e. look up field A.

There's obviously the question of reporting errors about using the code incorrectly back up to the caller... although I assume the usage on the VRL side may already be constrained anyways.

Reply:
> Just curious: was it not possible to coalesce this method between the `match_datadog_query` VRL function and this condition?

I think it would be. This change was intrusive enough that I didn't want to go too far afield from the basic idea but I do think more juice could be wrung from this basic line of work. 

> I know that a VRL value can be different from the more fixed layout of `LogEvent`, but it seems like -- and this is just from a cursory glance, to be fair -- that it would end being more about a simple translation layer to "query" the operand in a generic way i.e. look up field A.
> 
> There's obviously the question of reporting errors about using the code incorrectly back up to the caller... although I assume the usage on the VRL side may already be constrained anyways.

Yep, yep. Especially as the underlying path lookup code, value code unifies it seems natural to me that these two, uh, views into the same kind of search could unify.